### PR TITLE
mgr/dashboard: fix columns in host table  with NaN Undefined

### PR DESCRIPTION
**Pull request by avanthakkar**
Fixes: https://tracker.ceph.com/issues/54068
Signed-off-by: Avan Thakkar <athakkar@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by nizamial09** [comment]
jenkins test dashboard cephadm

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts [hunk 1]
```diff
@@ -464,8 +464,13 @@ export class HostsComponent extends ListWithDetails implements OnDestroy, OnInit
...
       _.forEach(this.hosts, (hostKey) => {
-        hostKey['memory_total_bytes'] = hostKey['memory_total_kb'] * 1024;
-        hostKey['raw_capacity'] = hostKey['hdd_capacity_bytes'] + hostKey['flash_capacity_bytes'];
+        hostKey['memory_total_bytes'] = !isNaN(hostKey['memory_total_kb'])
+          ? hostKey['memory_total_kb'] * 1024
+          : 'N/A';
+        hostKey['raw_capacity'] =
+          !isNaN(hostKey['hdd_capacity_bytes']) && !isNaN(hostKey['flash_capacity_bytes'])
+            ? hostKey['hdd_capacity_bytes'] + hostKey['flash_capacity_bytes']
+            : 'N/A';
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:473]: honestly, if a ternary assignment is not a one liner I would prefer a usual if statement.

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:473]: I avoided writing if else block twice)

**Review by nizamial09 — APPROVED** [review]

**Comment by avanthakkar** [comment]
jenkins test make check

**Review by pereman2 — APPROVED** [review]

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts [hunk 2]
```diff
@@ -464,8 +464,13 @@ export class HostsComponent extends ListWithDetails implements OnDestroy, OnInit
   transformHostsData() {
     if (this.checkHostsFactsAvailable()) {
       _.forEach(this.hosts, (hostKey) => {
-        hostKey['memory_total_bytes'] = hostKey['memory_total_kb'] * 1024;
-        hostKey['raw_capacity'] = hostKey['hdd_capacity_bytes'] + hostKey['flash_capacity_bytes'];
+        hostKey['memory_total_bytes'] = !isNaN(hostKey['memory_total_kb'])
+          ? hostKey['memory_total_kb'] * 1024
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:468]: Why do we care about presentation stuff in the logic layer? `NaN` or `"N/A"` is not relevant at this stage.

IMHO that should be left for the template and pipes are good at this. We even have a pipe `empty` that basically converts `null` and `undefined` to `"-"`. If we need to deal with `NaN` let's add that check to that pipe and keep a consistent behaviour ("-" for null/undefined in some places, but "N/A" in others). That's the good thing of reusing helpers.

And going one step further: if this `NaN/null/undefined` seems to be a recurrent issue (I remember at least another 2 recent PRs dealing with it), why don't we append the `empty` pipe by default to the `cd-datatable` cells? That way we'll avoid having to manual check for this in every single piece of data that we display.

Additionally, is it normal to get a `NaN` here? Shouldn't we get a null or undefined?

**Review by epuertat — CHANGES_REQUESTED** [review]

**Comment by epuertat** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test windows

**Comment by avanthakkar** [comment]
jenkins test windows

**Comment by epuertat** [comment]
jenkins test windows

**Pull request merged**